appengine/genetics/js/cage.js

```diff
@@ -0,0 +1,691 @@
+    Genetics.Cage.end('NONE_LEFT');
+    return;
+  }
+  // Check which genes have won.
+  var pickFightWinner = Genetics.Cage.aliveMice_[0].pickFightOwner;
+  var mateQuestionWinner = Genetics.Cage.aliveMice_[0].proposeMateOwner;
+  var acceptMateWinner = Genetics.Cage.aliveMice_[0].acceptMateOwner;
+  for (var i = 1; i < Genetics.Cage.aliveMice_.length; i++) {
+    var mouse = Genetics.Cage.aliveMice_[i];
+    if (pickFightWinner && pickFightWinner != mouse.pickFightOwner) {
+      pickFightWinner = null;
+    }
+    if (mateQuestionWinner && mateQuestionWinner != mouse.proposeMateOwner) {
+      mateQuestionWinner = null;
+    }
+    if (acceptMateWinner && acceptMateWinner != mouse.acceptMateOwner) {
+      acceptMateWinner = null;
+    }
+    if (!pickFightWinner && !mateQuestionWinner && !acceptMateWinner) {
+      break;
+    }
+  }
+  // End the game if there is a winner for all functions.
+  if (pickFightWinner && mateQuestionWinner && acceptMateWinner) {
+    Genetics.Cage.end('DOMINATION', pickFightWinner, mateQuestionWinner,
+        acceptMateWinner);
+  }
+  // Check if game has gone on too long.
+  if (Date.now() > Genetics.Cage.endTime_) {
+    Genetics.Cage.end('TIMEOUT', pickFightWinner, mateQuestionWinner,
+        acceptMateWinner);
+  }
+};
+
+/**
+ * Clears queued events for end of simulation.
+ * @param {string} cause The cause of game end.
+ * @param {string=} opt_pickFightWinner The owner of all the pickFight functions
+ * in alive mice.
+ * @param {string=} opt_proposeMateWinner The owner of all the proposeMate
+ * functions in alive mice.
+ * @param {string=} opt_acceptMateWinner The owner of all the acceptMate
+ * functions in alive mice.
+ */
+Genetics.Cage.end = function(cause, opt_pickFightWinner, opt_proposeMateWinner,
+    opt_acceptMateWinner) {
+  // Cancel all queued life simulations.
+  for (var mouseId in Genetics.Cage.lifePidsMap_) {
+    clearTimeout(Genetics.Cage.lifePidsMap_[mouseId]);
+  }
+  var pickFightWinner = opt_pickFightWinner || 'NONE';
+  var proposeMateWinner = opt_proposeMateWinner || 'NONE';
+  var acceptMateWinner = opt_acceptMateWinner || 'NONE';
+  new Genetics.Cage.Event('END_GAME', cause, pickFightWinner, proposeMateWinner,
+      acceptMateWinner).addToQueue();
+};
+
+/**
+ * Starts up a new interpreter and gets the return value of the function.
+ * @param {!Genetics.Mouse} mouse The mouse to get the code from.
+ * @param {string} mouseFunction The function call to evaluate.
+ * @param {Genetics.Mouse=} opt_param The mouse passed as a parameter to the
+ * function.
+ * @return {Array.<*>}
+ */
+Genetics.Cage.runMouseFunction = function(mouse, mouseFunction, opt_param) {
+  // Get a new interpreter to run the player code.
+  var interpreter = Genetics.Cage.getInterpreter(mouse, mouseFunction,
+      opt_param);
+  // Try running the player code.
+  try {
+    var ticks = Genetics.Cage.FUNCTION_TIMEOUT_STEPS;
+    while (interpreter.step()) {
+      if (ticks-- == 0) {
```

Comment:
This case is safe, I think.  However, I've learned the hard way that it is much better to use (ticks-- <= 0) since otherwise if by some bug it slips negative, the current code will run forever.